### PR TITLE
validate ipset set and entry - ipvs proxier validate entry before adding it and validate set before creating it

### LINKED ISSUE #57269: Should validate ipset util
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

There is no validation for ipset set and entry in pkg/util/ipset.

**What you expected to happen**:

We should validate them before call ipset exec. IPVS should validate ipset entry before adding it.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig network

/area ipvs

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

validate ipset util.

**Which issue(s) this PR fixes**:
Fixes #57269

**Special notes for your reviewer**:

I splitted this PR into some small pieces for the sake of code review.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by m1093782566** [comment]
cc @dnardo

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1104,6 +1104,10 @@ func (proxier *Proxier) syncProxyRules() {
 				IP2:      epIP,
 				SetType:  utilipset.HashIPPortIP,
 			}
+			if valid, err := proxier.loopbackSet.validateEntry(entry); !valid {
+				glog.Errorf("Failed to add entry %v to set %s since it's invalid, error: %v", entry, proxier.loopbackSet.Name, err)
```
Comment (brendandburns) [inline pkg/proxy/ipvs/proxier.go]: nit: given that this string is constant (and copied throughout) perhaps make it a constant in this file?

File: pkg/util/ipset/ipset.go [hunk 2]
```diff
@@ -309,17 +423,85 @@ func getIPSetVersionString(exec utilexec.Interface) (string, error) {
...
+	for _, valid := range ValidIPSetTypes {
+		if set == valid {
+			return true, nil
+		}
+	}
+	return false, fmt.Errorf("unsupported set type: %s", set)
+}
+
+// checks if given hash family is supported in ipset
+func validateHashFamily(family string) (bool, error) {
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: I don't see why this function should return an error...

File: pkg/util/ipset/ipset.go [hunk 3]
```diff
@@ -309,17 +423,85 @@ func getIPSetVersionString(exec utilexec.Interface) (string, error) {
...
+	if family == ProtocolFamilyIPV4 || family == ProtocolFamilyIPV6 {
+		return true, nil
+	}
+	return false, fmt.Errorf("unsupported hash family: %s", family)
+}
+
+// checks if given hash family is supported in ipset
+func validateProtocol(protocol string) (bool, error) {
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: I don't see why this function returns an error...

File: pkg/util/ipset/ipset.go [hunk 4]
```diff
@@ -309,17 +423,85 @@ func getIPSetVersionString(exec utilexec.Interface) (string, error) {
...
+		}
+		if num < 0 {
+			return false, fmt.Errorf("port number %d should be >=0", num)
+		}
+	}
+	return true, nil
+}
+
+// checks if the given ipset type is valid.
+func validateIPSetType(set Type) (bool, error) {
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: I don't see why this function returns an error...

File: pkg/util/ipset/ipset.go [hunk 5]
```diff
@@ -309,17 +423,85 @@ func getIPSetVersionString(exec utilexec.Interface) (string, error) {
...
+	strs := strings.Split(portRange, "-")
+	if len(strs) != 2 {
+		// port number -1 indicates invalid
+		return -1, -1, fmt.Errorf("port range should be in the format of `a-b`")
+	}
+	for i := range strs {
+		num, err := strconv.Atoi(strs[i])
+		if err != nil {
+			// port number -1 indicates invalid
+			return -1, -1, fmt.Errorf("%s is not an integer number", strs[i])
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: don't wrap `err` just return it...

File: pkg/util/ipset/ipset.go [hunk 6]
```diff
@@ -309,17 +423,85 @@ func getIPSetVersionString(exec utilexec.Interface) (string, error) {
...
 	if len(strs) != 2 {
-		return false
+		return false, fmt.Errorf("port range should be in the format of `a-b`")
 	}
 	for i := range strs {
-		if _, err := strconv.Atoi(strs[i]); err != nil {
-			return false
+		num, err := strconv.Atoi(strs[i])
+		if err != nil {
+			return false, fmt.Errorf("%s is not an integer number", strs[i])
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: don't wrap `err`

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Sounds fair, I will try. However, the error string contains some variables and the variables will break the complete string. For instance,

```
const EntryInvalidErr = "entry is invalid"

glog.Error("Failed to add entry %v to set %s, %s, error: %v", entry, set, EntryInvalidErr, err)

```

File: pkg/util/ipset/ipset.go (hunk 2 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: I return an error just for the sake of telling caller the reason... 

Do you think logging the error is a good idea?

File: pkg/util/ipset/ipset.go (hunk 4 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: I return an error just for the sake of telling caller the reason...

Do you think logging the error is a good idea?

File: pkg/util/ipset/ipset.go (hunk 6 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Okay, it's fixed now. PTAL.

File: pkg/util/ipset/ipset.go (hunk 4 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Perhaps we can return the error in caller side. 

It's fixed now, PTAL.

File: pkg/util/ipset/ipset.go (hunk 5 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Okay, it's fixed now. PTAL.

File: pkg/util/ipset/ipset.go (hunk 3 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: It's fixed now. PTAL.

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Applied, PTAL.

**Comment by m1093782566** [comment]
@brendandburns All comments are addressed now. Please take a look again when you have time.

File: pkg/proxy/ipvs/proxier.go [hunk 7]
```diff
@@ -1104,6 +1107,10 @@ func (proxier *Proxier) syncProxyRules() {
 				IP2:      epIP,
 				SetType:  utilipset.HashIPPortIP,
 			}
+			if valid, err := proxier.loopbackSet.validateEntry(entry); !valid {
+				glog.Errorf("Failed to add entry %v to set %s, error: %s, %v", entry, proxier.loopbackSet.Name, EntryInvalidErr, err)
```
Comment (brendandburns) [inline pkg/proxy/ipvs/proxier.go]: shouldnt we extract this string as a const?

**Comment by brendandburns** [comment]
i still think its weird for validation functions to return errors,

**Comment by m1093782566** [comment]
> i still think its weird for validation functions to return errors,

I can find many examples, 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/resource/get.go#L212

https://github.com/kubernetes/kubernetes/blob/master/pkg/security/podsecuritypolicy/selinux/types.go#L29

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/events/event/strategy.go#L58

I agree return both `bool` and `error` seems weird, and I think we need the error message to caller why it's invalid. So, what about just return error and if error occurs then it means invalid?

File: pkg/proxy/ipvs/proxier.go (hunk 7 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Because there are some variables in the string, e.g. $entry, $set, &err

What about:

```
const EntryInvalidErr = "error adding entry to ipset since entry is invalid"

glog.Errorf("%s, entry: %s, set: %s, reason: %s", EntryInvalidErr, entry, proxier.loopbackSet.Name, err)
```

Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]:
```
const EntryInvalidErr = "error adding entry to set since it's invalid"

glog.Errorf("%s", EntryInvalidErr)
```

**Comment by m1093782566** [comment]
@brendandburns 

All comments are applied now, please to see if you are happy with these fixes :)

**Comment by m1093782566** [comment]
ping @brendandburns ~

File: pkg/proxy/ipvs/proxier.go [hunk 8]
```diff
@@ -1144,8 +1151,16 @@ func (proxier *Proxier) syncProxyRules() {
 		// proxier.kubeServiceAccessSet.activeEntries.Insert(entry.String())
 		// Install masquerade rules if 'masqueradeAll' or 'clusterCIDR' is specified.
 		if proxier.masqueradeAll {
+			if valid := proxier.clusterIPSet.validateEntry(entry); !valid {
```
Comment (brendandburns) [inline pkg/proxy/ipvs/proxier.go]: This code is identical in both cases, move it up above the if/else conditional.

File: pkg/proxy/ipvs/proxier.go [hunk 9]
```diff
@@ -1352,8 +1387,16 @@ func (proxier *Proxier) syncProxyRules() {
 				}
 				switch protocol {
 				case "tcp":
+					if valid := proxier.nodePortSetTCP.validateEntry(entry); !valid {
```
Comment (brendandburns) [inline pkg/proxy/ipvs/proxier.go]: This can be done tighter/cleaner with:

var string name
switch protocol {
   case "tcp":
      name = ...
...
}
if valid := .... {
...
}
```

File: pkg/util/ipset/ipset.go [hunk 10]
```diff
@@ -102,10 +138,108 @@ type Entry struct {
...
+		if net.ParseIP(e.IP) == nil {
+			glog.Errorf("Error parsing entry %v ip address %v for ipset %v", e, e.IP, set)
+			return false
+		}
+
+		if valid := validateProtocol(e.Protocol); !valid {
+			return false
+		}
+
+		if e.Port < 0 {
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: A bunch of these checks are identical for all set types, pull them up and don't repeat the code.

File: pkg/proxy/ipvs/proxier.go (hunk 8 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Done

File: pkg/proxy/ipvs/proxier.go (hunk 9 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Done.

File: pkg/util/ipset/ipset.go (hunk 10 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Done.

**Comment by m1093782566** [comment]
@brendandburns 

Glad to tell you that all comments are addressed now. PTAL.

**Comment by m1093782566** [comment]
ping @brendandburns ~

File: pkg/util/ipset/OWNERS [hunk 11]
```diff
@@ -0,0 +1,8 @@
+approvers:
+- thockin
+- brendandburns
+- m1093782566
```
Comment (brendandburns) [inline pkg/util/ipset/OWNERS]: Let's do OWNERS files updates in a separate PR.

File: pkg/util/ipset/ipset.go [hunk 12]
```diff
@@ -341,4 +484,49 @@ func IsNotFoundError(err error) bool {
...
+	if protocol == ProtocolTCP || protocol == ProtocolUDP {
+		return true
+	}
+	glog.Errorf("Invalid entry's protocol: %s, supported protocols are [%s, %s]", protocol, ProtocolTCP, ProtocolUDP)
+	return false
+}
+
+// parsePortRange parse the begin and end port from a raw string(format: a-b).  beginPort <= endPort
+// in the return value.
+func parsePortRange(portRange string) (beginPort int, endPort int, err error) {
```
Comment (brendandburns) [inline pkg/util/ipset/ipset.go]: This code looks semi-fiddly, is there a unit test somewhere? I don't see it in this PR.

**Comment by brendandburns** [comment]
Sorry for the delay. A few more comments, mainly break out the OWNERS file into a separate PR and a request for some more unit tests.

Please ping me on Slack or mail if I'm lagging more than 2-3 days on the review.

Thanks

File: pkg/util/ipset/ipset.go (hunk 12 above)
Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Yes, I already added a unit test for this function. Please check: https://github.com/kubernetes/kubernetes/pull/57268/files#diff-1e6b0bd1e6429733c0d1633fe4741a98R799

File: pkg/util/ipset/OWNERS (hunk 11 above)
Comment (m1093782566) [reply pkg/util/ipset/OWNERS]: Another PR is already sent: https://github.com/kubernetes/kubernetes/pull/57582

**Comment by m1093782566** [comment]
It's fixed now. PTAL, thanks!

**Comment by m1093782566** [comment]
@kubernetes/sig-network-pr-reviews

**Comment by m1093782566** [comment]
@cmluciano @caseydavenport would you please help review/approve this PR?

**Comment by cmluciano** [comment]
cc @thockin @matchstick for approval

File: pkg/util/ipset/OWNERS [hunk 13]
```diff
@@ -0,0 +1,8 @@
+reviewers:
```
Comment (rramkumar1) [inline pkg/util/ipset/OWNERS]: This file was pushed in #57582. Can you remove here to keep this PR a little cleaner?

File: pkg/util/ipset/ipset.go [hunk 14]
```diff
@@ -85,6 +88,39 @@ type IPSet struct {
...
+		}
+	}
+	// check hash size value of ipset
+	if set.HashSize <= 0 {
+
+		return false
+	}
+	// check max elem value of ipset
+	if set.MaxElem <= 0 {
+		glog.Errorf("Invalid maxelem value %d, should be >0", set.MaxElem)
```
Comment (rramkumar1) [inline pkg/util/ipset/ipset.go]: Can you make the error "%d is an invalid max value for set, should be > 0"

File: pkg/proxy/ipvs/proxier.go [hunk 15]
```diff
@@ -975,6 +975,9 @@ func (proxier *Proxier) OnEndpointsSynced() {
 	proxier.syncProxyRules()
 }
 
+// EntryInvalidErr indiates if an ipset entry is invalid or not
```
Comment (rramkumar1) [inline pkg/proxy/ipvs/proxier.go]: Nit: "indicates"

File: pkg/proxy/ipvs/proxier.go [hunk 16]
```diff
@@ -975,6 +975,9 @@ func (proxier *Proxier) OnEndpointsSynced() {
...
+const EntryInvalidErr = "error adding entry %s to ipset %s since entry is invalid"
```
Comment (rramkumar1) [inline pkg/proxy/ipvs/proxier.go]: Can you reduce the error down to maybe "error adding entry %s to ipset %s". The part "since entry is invalid" That is a bit redundant.

File: pkg/util/ipset/OWNERS (hunk 13 above)
Comment (m1093782566) [reply pkg/util/ipset/OWNERS]: Sure. I rebased to the HEAD. Thanks!

File: pkg/proxy/ipvs/proxier.go (hunk 16 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Yeah, it's cleaner. It's fixed now, thanks!

File: pkg/proxy/ipvs/proxier.go (hunk 15 above)
Comment (m1093782566) [reply pkg/proxy/ipvs/proxier.go]: Done.

**Comment by m1093782566** [comment]
@brendandburns @rramkumar1 

Comments are addressed now. PTAL, THANKS!

File: pkg/util/ipset/ipset.go [hunk 17]
```diff
@@ -48,6 +50,7 @@ type Interface interface {
 	ListSets() ([]string, error)
 	// GetVersion returns the "X.Y" version string for ipset.
 	GetVersion() (string, error)
+	// TODO: add comment message for ipset
```
Comment (rramkumar1) [inline pkg/util/ipset/ipset.go]: I dont understand this TODO. What are you referring to?

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: All set types support the optional comment extension. Enabling this extension on an ipset enables you to annotate an ipset entry with an arbitrary string. This string is completely ignored by both the kernel and ipset itself and is purely for providing a convenient means to document the reason for an entry's existence. 

For example, the following shell command is illegal:
ipset add foo 1.1.1.1 comment `this comment is for something`

I just came up an idea about that but I am not sure if we really need it.

I can remove this line for now and will fire another PR if needed.

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: I just pushed an update to remove this TODO.

File: pkg/util/ipset/ipset.go [hunk 18]
```diff
@@ -85,6 +87,39 @@ type IPSet struct {
 	MaxElem int
 	// PortRange specifies the port range of bitmap:port type ipset.
 	PortRange string
+	// TODO: add comment message for ipset
+}
+
+// Validate checks if a given ipset is valid or not.
+func (set *IPSet) Validate() bool {
```
Comment (rramkumar1) [inline pkg/util/ipset/ipset.go]: I'm a little confused on why you need this. It seems like Validate for an Entry does everything you need. Can you describe a chain of events where this function would return false?

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: This function is called by 

```
func (runner *runner) CreateSet(set *IPSet, ignoreExistErr bool) error {
    ...
    // Validate ipset before creating
    valid := set.Validate()
    if !valid {
	return fmt.Errorf("Error creating ipset since it's invalid")
    }
    return runner.createSet(set, ignoreExistErr)
}
```

Possibly I need to change the PR title to `validate ipset set and entry - ipvs proxier validate entry before adding it and validate set before creating it`

Comment (rramkumar1) [reply pkg/util/ipset/ipset.go]: What I was asking was when would there be a situation where Validate() returns false? In other words, is there a set of consecutive events that would cause a set to be invalid? I can't think of any which is why I am confused as to why you even need it.

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: Yes, three are...

```
// Check if protocol is valid for `HashIPPort`, `HashIPPortIP` and `HashIPPortNet` type set.
	if set.SetType == HashIPPort || set.SetType == HashIPPortIP || set.SetType == HashIPPortNet {
		if valid := validateHashFamily(set.HashFamily); !valid {
			return false
		}
	}
	// check set type
	if valid := validateIPSetType(set.SetType); !valid {
		return false
	}
	// check port range for bitmap type set
	if set.SetType == BitmapPort {
		if valid := validatePortRange(set.PortRange); !valid {
			return false
		}
	}
	// check hash size value of ipset
	if set.HashSize <= 0 {
		return false
	}
	// check max elem value of ipset
	if set.MaxElem <= 0 {
		glog.Errorf("Invalid maxelem value %d, should be >0", set.MaxElem)
		return false
	}
```

All situations above will return false by Validate().

I am not sure I FULLY understand...

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: We should keep it in mind that this util pkg is public and can be used by other processes...

Comment (rramkumar1) [reply pkg/util/ipset/ipset.go]: I get that since it is a public pkg, you need it. However I am trying to understand at a deeper level why an IPSet would ever be invalid in the ipvs case

Comment (m1093782566) [reply pkg/util/ipset/ipset.go]: >  However I am trying to understand at a deeper level why an IPSet would ever be invalid in the ipvs case

Ah, ipvs case will not cause error for now.

Comment (rramkumar1) [reply pkg/util/ipset/ipset.go]: Got it, thank you for clarifying :)
